..Index: qt3-3.5.0/src/styles/qcommonstyle.cpp
===================================================================
--- src/styles/qcommonstyle.cpp
+++ src/styles/qcommonstyle.cpp
@@ -137,7 +137,7 @@ static const char * const check_list_con
 "                ",
 "                "};
 
-#include <nqmetaobject.h>
+#include <qmetaobject.h>
 
 QStringList getObjectTypeListForObject(const QObject* object) {
 	QStringList objectTypes;
