--- src/sql/qt_sql.pri	2011-10-30 19:55:58.000000000 +0100
+++ src/sql/qt_sql.pri	2008-01-15 20:09:13.000000000 +0100
@@ -116,7 +116,7 @@
 
 		unix {
 			!contains( LIBS, .*odbc.* ) {
-				LIBS        *= -liodbc
+				LIBS        *= -lodbc
 			}
 		}
 
--- plugins/src/sqldrivers/odbc/odbc.pro	2011-10-30 19:55:57.000000000 +0100
+++ plugins/src/sqldrivers/odbc/odbc.pro	2008-01-15 20:09:17.000000000 +0100
@@ -17,7 +17,7 @@
 unix {
 	OBJECTS_DIR	= .obj
 	!contains( LIBS, .*odbc.* ) {
-	    LIBS 	*= -liodbc
+	    LIBS 	*= -lodbc
 	}
 }
 
