Index: khtml/html/html_elementimpl.cpp
===================================================================
--- khtml/html/html_elementimpl.cpp.orig
+++ khtml/html/html_elementimpl.cpp
@@ -565,6 +565,13 @@ DocumentFragment HTMLElementImpl::create
 
 void HTMLElementImpl::setInnerHTML( const DOMString &html, int &exceptioncode )
 {
+ // Works line innerText in Gecko
+ // ### test if needed for ID_SCRIPT as well.
+ if ( id() == ID_STYLE ) {
+ setInnerText(html, exceptioncode);
+ return;
+ }
+
     DocumentFragment fragment = createContextualFragment( html );
     if ( fragment.isNull() ) {
         exceptioncode = DOMException::NO_MODIFICATION_ALLOWED_ERR;
