Index: kdm/backend/client.c
===================================================================
--- kdm/backend/client.c.orig
+++ kdm/backend/client.c
@@ -386,6 +386,9 @@ AccNoPass( const char *un, struct passwd
 	if (cursource != PWSRC_MANUAL)
 		return 1;
 
+	if (td->noPassAllUsers)
+		return 1;
+
 	for (hg = 0, fp = td->noPassUsers; *fp; fp++)
 		if (**fp == '@')
 			hg = 1;
Index: kdm/config.def
===================================================================
--- kdm/config.def.orig
+++ kdm/config.def
@@ -1852,6 +1852,15 @@ Description:
  (and any other user with UID = 0).
  <emphasis>Never</emphasis> list <systemitem class="username">root</systemitem>.
 
+Key: NoPassAllUsers
+Type: bool
+Default: false
+User: core
+Instance: #:0/true
+Comment: &
+Description:
+ All users can login without password
+
 Key: AutoLoginEnable
 Type: bool
 Default: false
