Index: kdm/backend/Makefile.am
===================================================================
--- kdm/backend/Makefile.am.orig
+++ kdm/backend/Makefile.am
@@ -1,11 +1,6 @@
 # forcibly remove thread-related defines & flags
 AUTOMAKE_OPTIONS = foreign
-CFLAGS = $(XDM_CFLAGS) -fno-strict-aliasing
-CPPFLAGS = $(USER_INCLUDES) $(X_INCLUDES) $(KRB4_INCS) $(KRB5_INCS) -I.. -I../..
-LDFLAGS = $(USER_LDFLAGS) $(X_LDFLAGS) $(X_RPATH) $(KRB4_RPATH) $(KRB5_RPATH)
-LDADD = $(LIB_X11) -lXau $(LIBXDMCP) $(PASSWDLIBS) $(LIBSHADOW) $(LIBGEN) \
-    $(LIB_LIBS) $(KRB4_LIBS) $(KRB5_LIBS) $(LIBSOCKET) $(LIBRESOLV) \
-    $(LIBUCB) $(LIBUTIL) $(LIBPOSIX4)
+AM_CPPFLAGS = $(USER_INCLUDES) $(X_INCLUDES) $(KRB4_INCS) $(KRB5_INCS) -I.. -I../..
 
 bin_PROGRAMS = kdm
 kdm_SOURCES = \
@@ -39,6 +34,11 @@ kdm_SOURCES = \
     xdmauth.c \
     xdmcp.c
 
+kdm_LDFLAGS = $(USER_LDFLAGS) $(X_LDFLAGS) $(X_RPATH) $(KRB4_RPATH) $(KRB5_RPATH)
+kdm_LDADD = $(LIB_X11) -lXau $(LIBXDMCP) $(PASSWDLIBS) \
+    $(LIB_LIBS) $(KRB4_LIBS) $(KRB5_LIBS) $(LIBSOCKET) $(LIBRESOLV) \
+    $(LIBUCB) $(LIBUTIL) $(LIBPOSIX4)
+
 EXTRA_DIST = printf.c
 
 noinst_HEADERS = dm.h dm_socket.h dm_error.h dm_auth.h greet.h
