Index: tdesu/tdesu/sudlg.cpp
===================================================================
--- tdesu/tdesu/sudlg.cpp.orig
+++ tdesu/tdesu/sudlg.cpp
@@ -30,18 +30,13 @@ KDEsuDialog::KDEsuDialog(TQCString user,
 
     TQString prompt;
     if (superUserCommand == "sudo" && m_User == "root") {
-	    prompt = i18n("Please enter your password." );
+	prompt = i18n("Please enter your password to continue.");
+    } else if (m_User == "root"){
+	prompt = i18n("Please enter the Administrator (root) password to continue.");
     } else {
-        if (m_User == "root") {
-	    prompt = i18n("The action you requested needs root privileges. "
-	    "Please enter root's password below or click "
-	    "Ignore to continue with your current privileges.");
-        } else {
-	    prompt = i18n("The action you requested needs additional privileges. "
-		"Please enter the password for \"%1\" below or click "
-		"Ignore to continue with your current privileges.").arg(static_cast<const char *>(m_User));
-	}
+	prompt = i18n("Please enter password for \"%1\" to continue.").arg(m_User);
     }
+    
     setPrompt(prompt);
     setKeepWarning(i18n("<qt>The stored password will be:<br> * Kept for up to %1 minutes<br> * Destroyed on logout").arg(timeout/60));
 
