Index: dbus/qdbusdatamap.h
===================================================================
--- dbus/qdbusdatamap.h	(Revision 785103)
+++ dbus/qdbusdatamap.h	(Revision 795238)
@@ -173,7 +173,7 @@
             }
             else if (hasContainerValueType())
             {
-                if (it.data()->buildSignature() != containerSignature)
+                if (it.data().buildDBusSignature() != containerSignature)
                 {
                     m_valueType = QDBusData::Invalid;
                     m_containerValueType = QDBusData();
Index: dbus/qdbuserror.h
===================================================================
--- dbus/qdbuserror.h	(Revision 785103)
+++ dbus/qdbuserror.h	(Revision 795238)
@@ -119,7 +119,7 @@
          * Returned by QDBusConnection's addConnection if the specified address
          * isn't a valid D-Bus bus address.
          *
-         * @see QDBusConnection:::addConnection(const QString&,const QString&);
+         * @see QDBusConnection::addConnection(const QString&,const QString&);
          */
         BadAddress,
 
Index: tools/dbusxml2qt3/classgen.cpp
===================================================================
--- tools/dbusxml2qt3/classgen.cpp	(Revision 785103)
+++ tools/dbusxml2qt3/classgen.cpp	(Revision 795238)
@@ -168,6 +168,17 @@
         extractForwardDeclarations(*it, forwards);
     }
 
+    it    = classData.signals.begin();
+    endIt = classData.signals.end();
+    for (; it != endIt; ++it)
+    {
+        if ((*it).arguments.isEmpty()) continue;
+
+        extractHeaderIncludes(*it, includes);
+        extractForwardDeclarations(*it, forwards);
+    }
+
+
     QValueList<Property>::const_iterator propertyIt = classData.properties.begin();
     QValueList<Property>::const_iterator propertyEndIt = classData.properties.end();
     for (; propertyIt != propertyEndIt; ++propertyIt)
Index: dbus-1-qt3.pc.in
===================================================================
--- dbus-1-qt3.pc.in	(Revision 785103)
+++ dbus-1-qt3.pc.in	(Revision 795238)
@@ -1,6 +1,6 @@
 prefix=@prefix@
 exec_prefix=@exec_prefix@
-libdir=${prefix}/lib
+libdir=@libdir@
 includedir=${prefix}/include/dbus-1.0/qt3
 
 Name: dbus-1-qt3
