Index: starttde
===================================================================
--- starttde.orig
+++ starttde
@@ -58,7 +58,7 @@ fi
 # people's heads. We use colours from the standard TDE palette for those with
 # palettised displays.
 if test -z "$XDM_MANAGED" || echo "$XDM_MANAGED" | grep ",auto" > /dev/null; then
-  xsetroot -solid "#618DCC" # sky blue
+  : # xsetroot -solid "#618DCC" # sky blue
 fi
 
 # Unset this for Darwin since it will screw up TDE's dynamic-loading.
@@ -588,7 +588,7 @@ echo "[starttde] Starting Trinity..." 1>
 $TDEDIR/bin/kdetcompmgr
 
 # Run KPersonalizer before the session if this is the first login.
-if test "$kpersonalizerrc_general_firstlogin" = "true"; then
+if false && test "$kpersonalizerrc_general_firstlogin" = "true"; then
   if [ ! -x $TDEDIR/bin/kpersonalizer ]; then
     echo "[starttde] kpersonalizer not found! Please install in order to properly configure your user profile." 1>&2
   else
@@ -618,6 +618,8 @@ if test -z "$dl"; then
   case "$ksplashrc_ksplash_theme" in
     None)
       ;; # Nothing.
+    ksplashx-*)
+      ;;
     Simple)
       if test "$kpersonalizerrc_general_firstlogin" = "true"; then
         $TDEDIR/bin/ksplashsimple
@@ -707,9 +709,11 @@ done
 echo "[starttde] Shutting down Trinity..." 1>&2
 
 # Clean up.
+if test -e $TDEDIR/bin/artsshell; then
+	$TDEDIR/bin/artsshell -q terminate
+fi
 $TDEDIR/bin/tdeinit_shutdown
 $TDEDIR/bin/dcopserver_shutdown --wait
-$TDEDIR/bin/artsshell -q terminate
 # KDE4 support.
 if [ -f /usr/bin/kdeinit4_shutdown ]; then
   /usr/bin/kde4 kdeinit4_shutdown 2>/dev/null
