Index: tools/assistant/lib/qassistantclient.cpp
===================================================================
--- tools/assistant/lib/qassistantclient.cpp.orig
+++ tools/assistant/lib/qassistantclient.cpp
@@ -135,7 +135,7 @@ TQAssistantClient::TQAssistantClient( co
     : TQObject( parent, name ), host ( "localhost" )
 {
     if ( path.isEmpty() )
-	assistantCommand = "assistant";
+	assistantCommand = "/usr/lib/qt3/bin/assistant";
     else {
 	TQFileInfo fi( path );
 	if ( fi.isDir() )
